sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -599,4 +617,240 @@ class ToNumberParser(numberFormat: String, errorOnFail: Boolean) extends Seriali
+   *                                decimal point
+   * @param inputAfterDecimalPoint string representation of the input decimal value after the
+   *                               decimal point
+   * @param reachedDecimalPoint true if we have reached the decimal point so far during processing
+   * @param result the result of formatting is built here as a string during iteration
+   */
+  private def formatDigitGroups(
+      groups: DigitGroups,
+      inputBeforeDecimalPoint: String,
+      inputAfterDecimalPoint: String,
+      reachedDecimalPoint: Boolean,
+      result: StringBuilder): Unit = {
+    // Iterate through the tokens in the DigitGroups. Reverse the order of the tokens so we
+    // consume them in the left-to-right order that they originally appeared in the format
+    // string.
+    for (digitGroupToken <- groups.tokens.reverse) {
+      digitGroupToken match {
+        case digits: Digits if !reachedDecimalPoint =>
+          val numDigits = digits match {
+            case ExactlyAsManyDigits(num) => num
+            case AtMostAsManyDigits(num) => num
+          }
+          for (_ <- 0 until numDigits) {
+            inputBeforeDecimalPoint(formattingBeforeDecimalPointIndex) match {
+              case c: Char if c != SPACE =>
+                result.append(c)
+              case _ => digits match {
+                case _: ExactlyAsManyDigits =>
+                  // The format string started with a 0 and had more digits than the provided
+                  // input string, so we prepend a 0 to the result.
+                  result.append(ZERO_DIGIT)
+                case _: AtMostAsManyDigits =>
+                  // The format string started with a 9 and had more digits than the provided
+                  // input string, so we prepend a space to the result.
+                  addSpaceCheckingOpenBracketOrMinusSign(result)
+              }
+            }
+            formattingBeforeDecimalPointIndex += 1
+          }
+        case digits: Digits if reachedDecimalPoint =>
+          val numDigits = digits match {
+            case ExactlyAsManyDigits(num) => num
+            case AtMostAsManyDigits(num) => num
+          }
+          for (_ <- 0 until numDigits) {
+            inputAfterDecimalPoint(formattingAfterDecimalPointIndex) match {
+              case c: Char if c != SPACE =>
+                result.append(c)
+              case _ =>
+                // The format string had more digits after the decimal point than the provided input
+                // string, so we append a space to the result.
+                result.append(SPACE)
+            }
+            formattingAfterDecimalPointIndex += 1
+          }
+        case _: ThousandsSeparator =>
+          if (result.nonEmpty && result.last >= ZERO_DIGIT && result.last <= NINE_DIGIT) {
+            result.append(COMMA_SIGN)
+          } else {
+            addSpaceCheckingOpenBracketOrMinusSign(result)
```

Comment:
This is also something we should document. If there is no enough digits in the input decimal, the thousands separator prints a space. Is it also the behavior in other databases?